server/pulp/server/async/tasks.py

```diff
@@ -373,20 +452,110 @@ def serialize(self):
+
+         See the docblock on _queue_reserved_task and _queue_reserved_task_list for more
+         information.
+
+         This method creates a TaskStatus as a placeholder for later updates. Pulp expects to poll
+         on a task just after calling this method, so a TaskStatus entry needs to exist for it
+         before it returns.
+
+         For a list of parameters accepted by the *args and **kwargs parameters, please see the
+         docblock for the apply_async() method.
+
+         :param reservation:    A string-or-list-of-strings that identify a set of named resources,
+                                guaranteeing that only one task reserving any resource-ids in this
+                                list can happen at a time.
+         :type  reservation:    list or basestr
+         :param tags:           A list of tags (strings) to place onto the task, used for searching
+                                for tasks by tag
+         :type  tags:           list
+         :param group_id:       The id to identify which group of tasks a task belongs to
+         :type  group_id:       uuid.UUID
+         :return:               An AsyncResult instance as returned by Celery's apply_async
+         :rtype:                celery.result.AsyncResult
+         """
+        inner_task_id = str(uuid.uuid4())
+        task_name = self.name
+        tag_list = kwargs.get('tags', [])
+        group_id = kwargs.get('group_id', None)
+
+        # Create a new task status with the task id and tags.
+        task_status = TaskStatus(task_id=inner_task_id, task_type=task_name,
+                                 state=constants.CALL_WAITING_STATE, tags=tag_list,
+                                 group_id=group_id)
+        # To avoid the race condition where __call__ method below is called before
+        # this change is propagated to all db nodes, using an 'upsert' here and setting
+        # the task state to 'waiting' only on an insert.
+        task_status.save_with_set_on_insert(fields_to_set_on_insert=['state', 'start_time'])
+        try:
+            # Decide what to call based on the kind-of reservation(s) we are being asked to make
+            if isinstance(reservation, list):
+                _queue_reserved_task_list.apply_async(
+                    args=[task_name, inner_task_id, reservation, args, kwargs],
+                    queue=RESOURCE_MANAGER_QUEUE
+                )
+            else:
+                _queue_reserved_task.apply_async(
+                    args=[task_name, inner_task_id, reservation, args, kwargs],
+                    queue=RESOURCE_MANAGER_QUEUE
+                )
+        except Exception:
+            TaskStatus.objects(task_id=task_status.task_id).update(state=constants.CALL_ERROR_STATE)
+            raise
+
+        return AsyncResult(inner_task_id)
+
+    def apply_async_with_reservation_list(self, resource_type, resource_id_list, *args, **kwargs):
```

Comment:
It would be better if we let the user specify a dictionary of resources where the type is the key and resource id is the value. What do you think about that?

Reply:
A couple of things 
- do you mean the/a key would be the type ('list' or 'str') and the value would be a list-of-resource-id-strs, or a single resource-id, depending on the type?
- do we want to handle the case where the dict has *both* list and str keys?
- Do we want even notice if the user passes in a type that is neither list nor str? (as written, if it's not-a list, we *assume* it's a string and safe to just pass along to the existing singler-resource method)

Consider that `reservation` could look like this:

`{ 'list': ['res_Id-1', 'res_id_2', 'res_id_3'], 'str': 'res_id_4', 'object': 'we-ignore-this'}`

In that case, we would have to 'notice' the list-and-str case, append the solo 'str' to the list, and then call `_queue_reserved_task_list` with the resulting combined list. (And whatever you passed in as not-list-or-str-type we would ignore. Or throw a "what are thinking?' exception?)

Reply:
`resources = {'repository': 'repo-id-foo', 'importer': 'importer-id-bar', 'repository': 'repo-id-shoe'}`

Reply:
@dkliban +1 to the simplicity of that, but I expected a list since it's has `_list` in the name.

Reply:
Maybe a list of tuples? [(resource_type, resource_id)]

Reply:
Concur w/list-of-tuples

Reply:
@dkliban @bmbouter tuple-accepting-API pushed as a new commit - feel free to throw moar rocks